substrate/frame/meta-tx/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Meta Tx
+//!
+//! TODO docs
+
+#![cfg_attr(not(feature = "std"), not_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+pub use pallet::*;
+
+use frame_support::{
+	dispatch::{DispatchInfo, GetDispatchInfo, PostDispatchInfo},
+	pallet_prelude::*,
+	traits::OriginTrait,
+};
+use frame_system::pallet_prelude::*;
+use sp_runtime::traits::TransactionExtension;
+use sp_runtime::traits::TransactionExtensionBase;
+use sp_runtime::traits::{Dispatchable, IdentifyAccount, Verify};
+
+/// Meta transaction type.
+// TODO: The `MetaTx` type is similar to `sp_runtime::generic::UncheckedExtrinsic`. However,
+// `MetaTx` cannot replace generic::UncheckedExtrinsic because we need to box the call type,
+// given that `MetaTx` is used as an argument type for a call.
+#[derive(Encode, Decode, PartialEq, Eq, TypeInfo, Clone, RuntimeDebug)]
+pub struct MetaTx<Address, Signature, Call, TxExtension> {
+	proof: Proof<Address, Signature>,
+	call: Box<Call>,
+	tx_ext: TxExtension,
```

Comment:
Is there any reason why we can't get away with just a `CheckNonce` here? The other stuff will be checked when the agent submits the tx anyway, we just want the replay protection for Alice.

If we want to move off the extension approach completely, we could do the nonce check and `inc` without the extension even, just run the `frame_system::Account` mutation inside the extrinsic, but the signed payload would need to include the nonce. Maybe it's too complicated, just a thought.